File: lib/api/adminContent.ts
import { AuthApiError } from "@/lib/api/auth";
import { apiUrl } from "@/lib/api/base";

export type AdminContentStatus = "draft" | "published" | "archived";
export type AdminDifficulty = "beginner" | "intermediate" | "advanced";

export interface AdminQuizOption {
  label: string;
  text: string;
}

export interface AdminQuizQuestion {
  id?: string;
  prompt: string;
  options: AdminQuizOption[];
  correctAnswer: string;
  explanation?: string;
  topic?: string;
  difficulty?: number;
  sourceDocument?: string;
  sourcePage?: number;
  irtDifficulty?: number;
  irtDiscrimination?: number;
  irtGuessing?: number;
  orderIndex: number;
  attemptCount?: number;
  averageScore?: number;
  discriminationIndex?: number;
}

export interface AdminQuizSummary {
  id: string;
  title: string;
  topic?: string;
  documentId?: string | null;
  difficulty?: string;
  questionCount: number;
  status: AdminContentStatus;
  usedBy: number;
  avgScore: number;
  estimatedMinutes: number;
  lastEditedAt?: string | null;
}

export interface AdminQuizDetail {
  id: string;
  title: string;
  description?: string | null;
  topic?: string | null;
  documentId?: string | null;
  difficulty?: string | null;
  estimatedMinutes: number;
  status: AdminContentStatus;
  questions: AdminQuizQuestion[];
}

export interface AdminFlashcardCard {
  id?: string;
  frontText: string;
  backText: string;
  topic?: string;
  difficulty?: number;
  sourceDocument?: string;
  sourcePage?: number;
  tags?: string[];
  orderIndex: number;
}

export interface AdminFlashcardDeckSummary {
  id: string;
  title: string;
  topic?: string | null;
  documentId?: string | null;
  cardCount: number;
  status: AdminContentStatus;
  usedBy: number;
  lastEditedAt?: string | null;
}

export interface AdminFlashcardDeckDetail {
  id: string;
  title: string;
  description?: string | null;
  topic?: string | null;
  documentId?: string | null;
  status: AdminContentStatus;
  cards: AdminFlashcardCard[];
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(apiUrl(path), {
    credentials: "include",
    cache: "no-store",
    ...init,
  });

  const contentType = response.headers.get("content-type") ?? "";
  const payload = contentType.includes("application/json")
    ? await response.json()
    : null;

  if (!response.ok) {
    const detail = payload?.detail;
    const message =
      typeof detail === "string"
        ? detail
        : detail?.message || payload?.message || "Request failed";
    throw new AuthApiError(message, response.status, detail);
  }

  return payload as T;
}

export async function listAdminQuizzes() {
  return request<AdminQuizSummary[]>("/admin/quizzes");
}

export async function getAdminQuiz(id: string) {
  return request<AdminQuizDetail>(`/admin/quizzes/${id}`);
}

export async function createAdminQuiz(input: Omit<AdminQuizDetail, "id">) {
  return request<AdminQuizDetail>("/admin/quizzes", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(input),
  });
}

export async function updateAdminQuiz(id: string, input: Omit<AdminQuizDetail, "id">) {
  return request<AdminQuizDetail>(`/admin/quizzes/${id}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(input),
  });
}

export async function publishAdminQuiz(id: string) {
  return request<AdminQuizDetail>(`/admin/quizzes/${id}/publish`, {
    method: "POST",
  });
}

export async function archiveAdminQuiz(id: string) {
  return request<AdminQuizDetail>(`/admin/quizzes/${id}/archive`, {
    method: "POST",
  });
}

export async function deleteAdminQuiz(id: string) {
  return request<void>(`/admin/quizzes/${id}`, {
    method: "DELETE",
  });
}

export async function listAdminFlashcardDecks() {
  return request<AdminFlashcardDeckSummary[]>("/admin/flashcard-decks");
}

export async function getAdminFlashcardDeck(id: string) {
  return request<AdminFlashcardDeckDetail>(`/admin/flashcard-decks/${id}`);
}

export async function createAdminFlashcardDeck(
  input: Omit<AdminFlashcardDeckDetail, "id">
) {
  return request<AdminFlashcardDeckDetail>("/admin/flashcard-decks", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(input),
  });
}

export async function updateAdminFlashcardDeck(
  id: string,
  input: Omit<AdminFlashcardDeckDetail, "id">
) {
  return request<AdminFlashcardDeckDetail>(`/admin/flashcard-decks/${id}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(input),
  });
}

export async function publishAdminFlashcardDeck(id: string) {
  return request<AdminFlashcardDeckDetail>(`/admin/flashcard-decks/${id}/publish`, {
    method: "POST",
  });
}

export async function archiveAdminFlashcardDeck(id: string) {
  return request<AdminFlashcardDeckDetail>(`/admin/flashcard-decks/${id}/archive`, {
    method: "POST",
  });
}

export async function deleteAdminFlashcardDeck(id: string) {
  return request<void>(`/admin/flashcard-decks/${id}`, {
    method: "DELETE",
  });
}

export interface AdminContentStats {
  questionPreview: string;
  attempts: number;
  avgScore: number;
  difficulty: string;
}

export async function getAdminContentStats(): Promise<AdminContentStats[]> {
  return request<AdminContentStats[]>("/admin/content/stats");
}

File: backend/app/schemas/admin_content.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models import ContentStatus, DifficultyLevel


class AdminQuizOption(BaseModel):
    label: str
    text: str


class AdminQuizQuestionPayload(BaseModel):
    id: str | None = None
    prompt: str = Field(min_length=2, max_length=4000)
    options: list[AdminQuizOption] = Field(min_length=2, max_length=6)
    correct_answer: str = Field(alias="correctAnswer", min_length=1, max_length=8)
    explanation: str | None = None
    source_document: str | None = Field(default=None, alias="sourceDocument", max_length=255)
    source_page: int | None = Field(default=None, alias="sourcePage", ge=1)
    irt_difficulty: int | None = Field(default=None, alias="irtDifficulty")
    irt_discrimination: float | None = Field(default=None, alias="irtDiscrimination")
    irt_guessing: float | None = Field(default=None, alias="irtGuessing")
    order_index: int = Field(default=0, alias="orderIndex", ge=0)

    model_config = ConfigDict(populate_by_name=True)


class AdminQuizPayload(BaseModel):
    title: str = Field(min_length=2, max_length=255)
    description: str | None = None
    topic: str | None = Field(default=None, max_length=128)
    difficulty: DifficultyLevel | None = None
    estimated_minutes: int = Field(default=10, alias="estimatedMinutes", ge=1, le=180)
    status: ContentStatus = ContentStatus.DRAFT
    questions: list[AdminQuizQuestionPayload] = Field(min_length=1)

    model_config = ConfigDict(populate_by_name=True)


class AdminQuizQuestionDetail(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    prompt: str
    options: list[AdminQuizOption]
    correct_answer: str = Field(alias="correctAnswer")
    explanation: str | None = None
    source_document: str | None = Field(default=None, alias="sourceDocument")
    source_page: int | None = Field(default=None, alias="sourcePage")
    irt_difficulty: int | None = Field(default=None, alias="irtDifficulty")
    irt_discrimination: float | None = Field(default=None, alias="irtDiscrimination")
    irt_guessing: float | None = Field(default=None, alias="irtGuessing")
    order_index: int = Field(alias="orderIndex")


class AdminQuizSummary(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    title: str
    topic: str | None = None
    difficulty: str | None = None
    question_count: int = Field(alias="questionCount")
    status: ContentStatus
    used_by: int = Field(alias="usedBy")
    avg_score: int = Field(alias="avgScore")
    estimated_minutes: int = Field(alias="estimatedMinutes")
    last_edited_at: datetime | None = Field(default=None, alias="lastEditedAt")


class AdminQuizDetail(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    title: str
    description: str | None = None
    topic: str | None = None
    difficulty: str | None = None
    estimated_minutes: int = Field(alias="estimatedMinutes")
    status: ContentStatus
    questions: list[AdminQuizQuestionDetail]


class AdminFlashcardPayload(BaseModel):
    id: str | None = None
    front_text: str = Field(alias="frontText", min_length=1, max_length=4000)
    back_text: str = Field(alias="backText", min_length=1, max_length=4000)
    source_document: str | None = Field(default=None, alias="sourceDocument", max_length=255)
    source_page: int | None = Field(default=None, alias="sourcePage", ge=1)
    tags: list[str] | None = Field(default=None)
    order_index: int = Field(default=0, alias="orderIndex", ge=0)

    model_config = ConfigDict(populate_by_name=True)


class AdminFlashcardDeckPayload(BaseModel):
    title: str = Field(min_length=2, max_length=255)
    description: str | None = None
    topic: str | None = Field(default=None, max_length=128)
    status: ContentStatus = ContentStatus.DRAFT
    cards: list[AdminFlashcardPayload] = Field(min_length=1)

    model_config = ConfigDict(populate_by_name=True)


class AdminFlashcardDetail(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    front_text: str = Field(alias="frontText")
    back_text: str = Field(alias="backText")
    source_document: str | None = Field(default=None, alias="sourceDocument")
    source_page: int | None = Field(default=None, alias="sourcePage")
    tags: list[str] | None = None
    order_index: int = Field(alias="orderIndex")


class AdminFlashcardDeckSummary(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    title: str
    topic: str | None = None
    card_count: int = Field(alias="cardCount")
    status: ContentStatus
    used_by: int = Field(alias="usedBy")
    last_edited_at: datetime | None = Field(default=None, alias="lastEditedAt")


class AdminFlashcardDeckDetail(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    title: str
    description: str | None = None
    topic: str | None = None
    status: ContentStatus
    cards: list[AdminFlashcardDetail]
